### PR TITLE
Implement ACLs on Digital Ocean

### PR DESCRIPTION
**Pull request by mmussomele**
This shouldn't be considered ready for merge, but it's at the point where it could be reviewed. Merge-ablility pending on running intergration tests.

### CODE REVIEW

File: cloud/digitalocean/digitalocean.go [hunk 1]
```diff
@@ -284,8 +316,102 @@ func (prvdr Provider) deleteAndWait(ids string) error {
...
+		}
+		curACLSet[key] = struct{}{}
+	}
+
+	var curACLs aclList
+	for key := range curACLSet {
+		curACLs = append(curACLs, key)
+	}
+
+	_, toAdd, toRemove := join.HashJoin(aclList(desired), curACLs, nil, nil)
```
Comment (kklin) [inline cloud/digitalocean/digitalocean.go]: There's an `acl.Slice` already

File: cloud/digitalocean/digitalocean.go [hunk 2]
```diff
@@ -284,8 +316,102 @@ func (prvdr Provider) deleteAndWait(ids string) error {
...
+			panic(err)
+		}
+
+		to, err := strconv.ParseInt(ports[1], 10, 0)
+		if err != nil {
+			panic(err)
+		}
+
+		key := acl.ACL{
+			CidrIP:  cur.Sources.Addresses[0],
```
Comment (kklin) [inline cloud/digitalocean/digitalocean.go]: I think we should just loop over all the addresses and generate ACLs for them. Then it's obvious that we aren't going to get an out of bounds error, or miss some sources.

File: cloud/digitalocean/digitalocean.go [hunk 3]
```diff
@@ -284,8 +316,102 @@ func (prvdr Provider) deleteAndWait(ids string) error {
...
+
+func syncACLs(desired []acl.ACL, current []godo.InboundRule) ([]godo.InboundRule,
+	[]godo.InboundRule) {
+
+	curACLSet := map[acl.ACL]struct{}{}
+	for _, cur := range current {
+		ports := strings.Split(cur.PortRange, "-")
+		from, err := strconv.ParseInt(ports[0], 10, 0)
+		if err != nil {
+			panic(err)
```
Comment (kklin) [inline cloud/digitalocean/digitalocean.go]: What if someone fiddled with the DigitalOcean UI and added some ACLs in a different format? I think we should just log here, and continue. We might even consider removing unparsable rules, not sure.

File: cloud/digitalocean/digitalocean.go [hunk 4]
```diff
@@ -284,8 +316,102 @@ func (prvdr Provider) deleteAndWait(ids string) error {
...
+		}
+
+		to, err := strconv.ParseInt(ports[1], 10, 0)
```
Comment (kklin) [inline cloud/digitalocean/digitalocean.go]: These `ParseInts` could just be `strconv.Atoi`

File: cloud/digitalocean/digitalocean.go (hunk 3 above)
Comment (kklin) [inline cloud/digitalocean/digitalocean.go]: A check on the length of `ports` would also be good I think

File: cloud/digitalocean/digitalocean.go [hunk 5]
```diff
@@ -284,8 +316,102 @@ func (prvdr Provider) deleteAndWait(ids string) error {
...
+	if _, err := prvdr.AddRules(firewall.ID, add); err != nil {
+		return err
+	}
+	if _, err := prvdr.RemoveRules(firewall.ID, remove); err != nil {
+		return err
+	}
 	return nil
 }
+
+func (prvdr Provider) getCreateFirewall() (*godo.Firewall, error) {
```
Comment (kklin) [inline cloud/digitalocean/digitalocean.go]: Any particular reason this should return a pointer?

File: cloud/digitalocean/digitalocean.go [hunk 6]
```diff
@@ -284,8 +316,102 @@ func (prvdr Provider) deleteAndWait(ids string) error {
...
+		if firewall.Name == prvdr.namespace {
+			return &firewall, nil
+		}
+	}
+
+	firewall, _, err := prvdr.CreateFirewall(prvdr.namespace, allowAll)
+	return firewall, err
+}
+
+func syncACLs(desired []acl.ACL, current []godo.InboundRule) ([]godo.InboundRule,
```
Comment (kklin) [inline cloud/digitalocean/digitalocean.go]: This function signature would be a little easier to read IMO if you named the return arguments

File: cloud/digitalocean/digitalocean.go [hunk 7]
```diff
@@ -176,24 +177,19 @@ func (prvdr Provider) List() (machines []db.Machine, err error) {
...
+	var eg errgroup.Group
 	for _, m := range bootSet {
 		if m.Preemptible {
 			return errors.New("preemptible instances are not yet implemented")
 		}
 
-		go func(m db.Machine) {
-			errChan <- prvdr.createAndAttach(m)
-		}(m)
+		m := m
```
Comment (kklin) [inline cloud/digitalocean/digitalocean.go]: This looks weird.. I could totally see someone removing that line in the future. I wonder if we should curry `m`:
```
func machineAction(m db.Machine, act func(db.Machine) error) func() error {
  return func() error {
    return act(m);
  }
}
```

Might be overkill though. At the least I'd name the variable something other than the iterator variable

File: cloud/digitalocean/client/mocks/Client.go [hunk 8]
```diff
@@ -1,5 +1,3 @@
-// Code generated by mockery v1.0.1 DO NOT EDIT.
```
Comment (kklin) [inline cloud/digitalocean/client/mocks/Client.go]: I don't think these headers should be removed

File: cloud/digitalocean/digitalocean.go [hunk 9]
```diff
@@ -34,6 +32,39 @@ var c = counter.New("Digital Ocean")
...
+// associated with it. It accepts the ones listed below, and we want to allow traffic
+// only based on IP and port, so allow them all.
+//
+// See https://developers.digitalocean.com/documentation/v2/#add-rules-to-a-firewall
+var protocols = []string{"tcp", "udp", "icmp"}
+
+// DigitalOcean firewalls block all traffic that is not explicitly allowed. We want to
+// allow all outgoing traffic.
+var allowAll = []godo.OutboundRule{
+	{
```
Comment (kklin) [inline cloud/digitalocean/digitalocean.go]: These are sort of repetitive, I wonder if we should generate them in a function. Don't feel strongly about it though.

File: cloud/digitalocean/digitalocean_test.go [hunk 10]
```diff
@@ -317,21 +317,119 @@ func TestStop(t *testing.T) {
...
 		{
-			CidrIP:  "ocean",
+			CidrIP:  "11.0.0.0/27",
 			MinPort: 22,
 			MaxPort: 22,
 		},
-	})
+	}
+
+	mc.On("CreateFirewall", testNamespace, allowAll).Return(
```
Comment (kklin) [inline cloud/digitalocean/digitalocean_test.go]: Could you add a comment above each test "group" saying the behavior you're testing? It'd make the test a lot easier to follow (I'll look at the tests after you've done this)

File: cloud/digitalocean/client/mocks/Client.go (hunk 8 above)
Comment (mmussomele) [inline cloud/digitalocean/client/mocks/Client.go]: hmmm odd I just ran `make generate`. Anyhow, I put them back

File: cloud/digitalocean/digitalocean.go (hunk 9 above)
Comment (mmussomele) [inline cloud/digitalocean/digitalocean.go]: I think leaving them is fine, but that reminds me I forgot IPv6, so I'll factor the destinations out and add that

File: cloud/digitalocean/digitalocean.go (hunk 5 above)
Comment (mmussomele) [inline cloud/digitalocean/digitalocean.go]: not really, changed

File: cloud/digitalocean/digitalocean.go (hunk 7 above)
Comment (mmussomele) [inline cloud/digitalocean/digitalocean.go]: I think wrapping it is worth it, especially if someone doesn't look carefully at why the assignment is there.

**Comment by mmussomele** [comment]
Ready for another look.

**Comment by kklin** [comment]
Sorry about the delay testing this!

I tried booting this branch and got errors of the form
```
WARNING [Sep  6 15:47:39.920] Could not update ACLs on DigitalOcean in sfo1. error=POST https://api.digitalocean.com/v2/firewalls: 422 (request "84cd4a2d-eaea-4162-b208-7db3e280877b") You must specify a positive value for ports.
```

I think it's failing to create the firewalls.

I know that you're busy with other things now @mmussomele, so unless you want to fix it up (which I know is tough since you don't have a DO token), one of us can take a look.

**Comment by kklin** [comment]
Closing because @aegamesi opened a new PR with some bug fixes

**Pull request closed without merging**